src/main/java/com/vmware/vipclient/i18n/messages/service/ComponentService.java

```diff
@@ -113,38 +110,52 @@ public MessageCacheItem getMessages(Iterator<Locale> fallbackLocalesIter) {
     		}
     	} else { // Item is not in cache.
 			ProductService ps = new ProductService(dto);
-			if (!ps.isSupportedLocale(Locale.forLanguageTag(dto.getLocale()))) { // Requested locale is not supported
-				Locale matchedLocale = LocaleUtility.pickupLocaleFromList(new LinkedList<>(ps.getSupportedLocales()), Locale.forLanguageTag(dto.getLocale()));
-				if (ps.isSupportedLocale(matchedLocale)) { // Requested locale matches a supported locale (eg. requested locale "fr_CA matches supported locale "fr")
+			Locale locale = Locale.forLanguageTag(dto.getLocale());
+			if (ps.isSupportedLocale(locale) || VIPCfg.getInstance().isPseudo()) {
+				cacheItem = createCacheItem(); // Fetch for the requested locale from data store, create cacheItem and store in cache
+				if (cacheItem.getCachedData().isEmpty())  // Failed to fetch messages for the requested locale
+					cacheItem = getFallbackLocaleMessages(fallbackLocalesIter);
+			} else { // Requested locale is not supported
+				Locale matchedLocale = LocaleUtility.pickupLocaleFromList(ps.getSupportedLocales(), locale);
+				if (matchedLocale != null) { // Requested locale matches a supported locale (eg. requested locale "fr_CA matches supported locale "fr")
 					MessagesDTO matchedLocaleDTO = new MessagesDTO(dto.getComponent(), matchedLocale.toLanguageTag(), dto.getProductID(), dto.getVersion());
-					return new ComponentService(matchedLocaleDTO).getMessages();
-				}
+					cacheItem = new ComponentService(matchedLocaleDTO).getMessages();
+				} else  // Requested locale is not supported and does not match any supported locales
+					cacheItem = getFallbackLocaleMessages(fallbackLocalesIter);
 			}
-
-			//Create and store cacheItem for the requested locale
-			cacheItem = createCacheItem(fallbackLocalesIter);
     	}
     	return cacheItem;
     }
 
+    /**
+     * Gets the messages in the next fallback locale by passing the next fallback locale DTO to a new instance of ComponentService
+	 * and then invoking {@link #getMessages(Iterator)}.
+	 * @param fallbackLocalesIter The fallback locale queue to use in case of failure. If null, no locale fallback will be applied.
+     */
+    private MessageCacheItem getFallbackLocaleMessages(Iterator<Locale> fallbackLocalesIter) {
+		if (fallbackLocalesIter != null && fallbackLocalesIter.hasNext()) {
+			Locale fallbackLocale = fallbackLocalesIter.next();
+			if (fallbackLocale.toLanguageTag().equals(dto.getLocale())) {
+				return getFallbackLocaleMessages(fallbackLocalesIter);
+			}
+			// Use MessageCacheItem of the next fallback locale.
+			MessagesDTO fallbackLocaleDTO = new MessagesDTO(dto.getComponent(), fallbackLocale.toLanguageTag(), dto.getProductID(), dto.getVersion());
+			return new ComponentService(fallbackLocaleDTO).getMessages(fallbackLocalesIter);
```

Comment:
In this way when dto locale is 'source' and source messages not found, it will fallback to default locale.

Reply:
If dto locale is source, it means it has reached the end of the fallback queue. Hence, it won't enter the if in line 136 because  `fallbackLocalesIter.hasNext()` is false.

Reply:
I mean when request locale is 'source', which make dto locale is 'source'

Reply:
Do you mean the preferred locale from the user is "source"? What use case do you have in mind where locale will be "source"? I suppose "source" will only be used as fallback.

Nonetheless, if user actually sends "source" as the preferred locale and source messages are not found, then yes, it should go though the usual locale fallback queue. What is your concern with this?

Reply:
Discussed this offline. Huihui is referring to this issue which is not within the scope of this PR: https://github.com/vmware/singleton/issues/839